internal/settings/providers/netcup/provider.go

```diff
@@ -0,0 +1,145 @@
+package netcup
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+
+	"github.com/qdm12/ddns-updater/internal/models"
+	"github.com/qdm12/ddns-updater/internal/settings/constants"
+	"github.com/qdm12/ddns-updater/internal/settings/errors"
+	"github.com/qdm12/ddns-updater/internal/settings/utils"
+	"github.com/qdm12/ddns-updater/pkg/publicip/ipversion"
+)
+
+type Provider struct {
+	customerNumber string
+	domain         string
+	host           string
+	ipVersion      ipversion.IPVersion
+	apiKey         string
+	password       string
+}
+
+func New(data json.RawMessage, domain, host string,
+	ipVersion ipversion.IPVersion) (p *Provider, err error) {
+	extraSettings := struct {
+		CustomerNumber string `json:"customer_number"`
+		APIKey         string `json:"api_key"`
+		Password       string `json:"password"`
+	}{}
+	if err := json.Unmarshal(data, &extraSettings); err != nil {
+		return nil, err
+	}
+	p = &Provider{
+		domain:         domain,
+		host:           host,
+		ipVersion:      ipVersion,
+		customerNumber: extraSettings.CustomerNumber,
+		apiKey:         extraSettings.APIKey,
+		password:       extraSettings.Password,
+	}
+	if err := p.isValid(); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func (p *Provider) isValid() error {
+	switch {
+	case p.customerNumber == "":
+		return errors.ErrEmptyCustomerNumber
+	case p.apiKey == "":
+		return errors.ErrEmptyAppKey
+	case p.password == "":
+		return errors.ErrEmptyPassword
+	case p.host == "*":
+		return errors.ErrHostWildcard
+	}
+	return nil
+}
+
+func (p *Provider) String() string {
+	return utils.ToString(p.domain, p.host, constants.Netcup, p.ipVersion)
+}
+
+func (p *Provider) Domain() string {
+	return p.domain
+}
+
+func (p *Provider) Host() string {
+	return p.host
+}
+
+func (p *Provider) IPVersion() ipversion.IPVersion {
+	return p.ipVersion
+}
+
+func (p *Provider) Proxied() bool {
+	return false
+}
+
+func (p *Provider) BuildDomainName() string {
+	return utils.BuildDomainName(p.host, p.domain)
+}
+
+func (p *Provider) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain:    models.HTML(fmt.Sprintf("<a href=\"http://%s\">%s</a>", p.BuildDomainName(), p.BuildDomainName())),
+		Host:      models.HTML(p.Host()),
+		Provider:  "<a href=\"https://www.netcup.eu/\">Netcup.eu</a>",
+		IPVersion: models.HTML(p.ipVersion.String()),
+	}
+}
+
+func (p *Provider) Update(ctx context.Context, client *http.Client, ip net.IP) (newIP net.IP, err error) {
+	u := url.URL{
+		Scheme:   "https",
+		Host:     "ccp.netcup.net",
+		Path:     "/run/webservice/servers/endpoint.php",
+		RawQuery: "JSON",
+	}
+	nc := NewClient(ctx, p.customerNumber, p.apiKey, p.password, u.String())
+
+	err = nc.Login(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	record, err := nc.GetRecordToUpdate(ctx, p.domain, p.host, ip)
```

Comment:
nit since we dereference record below, we should return a value instead of a pointer